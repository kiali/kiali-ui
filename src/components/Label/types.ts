export interface Labels {
  name: string;
  value: string;
}
