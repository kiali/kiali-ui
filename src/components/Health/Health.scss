svg:not(:root).icon-failure,
svg:not(:root).icon-degraded,
svg:not(:root).icon-healthy,
svg:not(:root).icon-na {
  overflow: visible;
}
.tippy-popper[x-placement^='left'] .pf-tooltip-theme .tippy-arrow {
  background-color: #fff;
  border-color: black;
  border-width: 1px;
  border-style: solid;
}
.health_indicator .pf-c-tooltip__content {
  background-color: #fff;
  border-style: solid;
  border-color: black;
  border-width: 1px;
  text-align: left;
}
