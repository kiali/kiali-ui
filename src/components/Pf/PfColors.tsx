export enum PfColors {
  Black100 = '#fafafa',
  Black150 = '#f5f5f5',
  Black200 = '#ededed',
  Black300 = '#d1d1d1',
  Black400 = '#bbb',
  Black500 = '#8b8d8f',
  Black600 = '#72767b',
  Black700 = '#4d5258',
  Black800 = '#393f44',
  Black900 = '#292e34',
  Black1000 = '#030303',
  Blue50 = '#def3ff',
  Blue100 = '#bee1f4',
  Blue200 = '#7dc3e8',
  Blue300 = '#39a5dc',
  Blue400 = '#0088ce',
  Blue500 = '#00659c',
  Blue600 = '#004368',
  Blue700 = '#002235',
  Cyan100 = '#bedee1',
  Cyan200 = '#7dbdc3',
  Cyan300 = '#3a9ca6',
  Cyan400 = '#007a87',
  Cyan500 = '#005c66',
  Cyan600 = '#003d44',
  Cyan700 = '#001f22',
  Gold100 = '#fbeabc',
  Gold200 = '#f9d67a',
  Gold300 = '#f5c12e',
  Gold400 = '#f0ab00',
  Gold500 = '#b58100',
  Gold600 = '#795600',
  Gold700 = '#3d2c00',
  Green100 = '#cfe7cd',
  Green200 = '#9ecf99',
  Green300 = '#6ec664',
  Green400 = '#3f9c35',
  Green500 = '#2d7623',
  Green600 = '#1e4f18',
  Green700 = '#0f280d',
  LightBlue100 = '#beedf9',
  LightBlue200 = '#7cdbf3',
  LightBlue300 = '#35caed',
  LightBlue400 = '#00b9e4',
  LightBlue500 = '#008bad',
  LightBlue600 = '#005c73',
  LightBlue700 = '#002d39',
  LightGreen100 = '#e4f5bc',
  LightGreen200 = '#c8eb79',
  LightGreen300 = '#ace12e',
  LightGreen400 = '#92d400',
  LightGreen500 = '#6ca100',
  LightGreen600 = '#486b00',
  LightGreen700 = '#253600',
  Orange100 = '#fbdebf',
  Orange200 = '#f7bd7f',
  Orange300 = '#f39d3c',
  Orange400 = '#ec7a08',
  Orange500 = '#b35c00',
  Orange600 = '#773d00',
  Orange700 = '#3b1f00',
  Purple100 = '#c7bfff',
  Purple200 = '#a18fff',
  Purple300 = '#8461f7',
  Purple400 = '#703fec',
  Purple500 = '#582fc0',
  Purple600 = '#40199a',
  Purple700 = '#1f0066',
  Red100 = '#cc0000',
  Red200 = '#a30000',
  Red300 = '#8b0000',
  Red400 = '#470000',
  Red500 = '#2c0000',

  White = '#fff',
  Black = '#030303',

  Blue = '#0088ce', // Blue400
  Cyan = '#007a87', // Cyan400
  Gold = '#f0ab00', // Gold400
  Green = '#3f9c35', // Green400
  LightBlue = '#00b9e4', // LightBlue400
  LightGreen = '#92d400', // LightGreen400
  Orange = '#ec7a08', // Orange400
  Red = '#cc0000', // Red100

  // Kiali colors that use PF colors
  Gray = '#72767b' // Black600
}

export enum PFColorVars {
  DangerColor = 'var(--pf-global--danger-color--100)',
  WarningColor = 'var(--pf-global--warning-color--100)',
  SuccessColor = 'var(--pf-global--success-color--100)'
}
