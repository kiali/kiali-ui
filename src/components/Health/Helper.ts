import * as React from 'react';
import { Status } from 'types/Health';

type Size = 'sm' | 'md' | 'lg' | 'xl';

export const createIcon = (status: Status, size?: Size) => {
  return React.createElement(status.icon, { color: status.color, size: size, className: status.class });
};
