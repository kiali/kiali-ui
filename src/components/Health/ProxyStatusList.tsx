import * as React from 'react';
import { ProxyStatus } from '../../types/Health';
import { Flex, FlexItem, Split, SplitItem, Text, TextVariants } from '@patternfly/react-core';
import { style } from 'typestyle';

type Props = {
  workloadName: string;
  statuses: ProxyStatus[];
};

const smallStyle = style({ fontSize: '70%' });
const workloadNameStyle = style({ minWidth: '6rem', overflowWrap: 'break-word' });

const proxyStatusSort = (a: ProxyStatus, b: ProxyStatus): number => {
  return a.component > b.component ? 1 : -1;
};

class ProxyStatusList extends React.Component<Props> {
  sortedStatus = () => {
    return this.props.statuses.sort(proxyStatusSort);
  };

  statusList = () => {
    const statusItems = this.sortedStatus().map((value: ProxyStatus, i: number) => {
      return (
        <FlexItem key={'proxy-status-' + i}>
          <Text component={TextVariants.small} className={smallStyle}>
            {value.component + ': ' + value.status}
          </Text>
        </FlexItem>
      );
    });

    return <Flex style={{ flex: 'inlineFlex' }}> {statusItems} </Flex>;
  };

  render() {
    if (this.props.statuses.length > 0) {
      return (
        <Split gutter={'sm'}>
          <SplitItem className={workloadNameStyle}>{this.props.workloadName}</SplitItem>
          <SplitItem>{this.statusList()}</SplitItem>
        </Split>
      );
    } else {
      return [];
    }
  }
}

export default ProxyStatusList;
