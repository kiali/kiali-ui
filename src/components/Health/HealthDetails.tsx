import * as React from 'react';
import * as H from '../../types/Health';
import { createIcon } from './Helper';
import { InfoAltIcon } from '@patternfly/react-icons';
import './Health.css';
import { PfColors } from '../Pf/PfColors';

interface Props {
  health: H.Health;
  tooltip?: boolean;
}

export class HealthDetails extends React.PureComponent<Props, {}> {
  renderErrorRate = (item: H.HealthItem, idx: number) => {
    const config = this.props.health.getStatusConfig();
    const isValueInConfig =
      config && this.props.health.health.statusConfig ? this.props.health.health.statusConfig.value > 0 : false;

    const showTraffic = item.children
      ? item.children.filter(sub => {
          const showItem = sub.value && sub.value > 0;
          return (sub.status !== H.HEALTHY && showItem) || !this.props.tooltip;
        }).length > 0
      : false;
    return showTraffic ? (
      <div key={idx}>
        <strong>
          {' ' + item.title + (item.text && item.text.length > 0 ? ': ' : '')}{' '}
          {config && <InfoAltIcon color={PfColors.Gray} />}
        </strong>
        {item.text}
        {item.children && (
          <ul style={{ listStyleType: 'none', paddingLeft: 12 }}>
            {item.children.map((sub, subIdx) => {
              const showItem = sub.value && sub.value > 0;
              return (sub.status !== H.HEALTHY && showItem) || !this.props.tooltip ? (
                <li key={subIdx}>
                  {createIcon(sub.status)} {sub.text}
                </li>
              ) : (
                <></>
              );
            })}
            {config && isValueInConfig && (
              <li key={'degraded_failure_config'}>
                {createIcon(H.DEGRADED)}: {config.degraded === 0 ? '>' : '>='}
                {config.degraded}% {createIcon(H.FAILURE)}: {config.degraded === 0 ? '>' : '>='}
                {config.failure}%
              </li>
            )}
          </ul>
        )}
      </div>
    ) : (
      <></>
    );
  };

  renderChildren = (item: H.HealthItem, idx: number) => {
    return item.title.startsWith(H.TRAFFICSTATUS)
      ? this.renderErrorRate(item, idx)
      : (item.status !== H.HEALTHY || !this.props.tooltip) && (
          <div key={idx}>
            <strong>{' ' + item.title + (item.text && item.text.length > 0 ? ': ' : '')}</strong>
            {item.text}
            {item.children && (
              <ul style={{ listStyleType: 'none', paddingLeft: 12 }}>
                {item.children.map((sub, subIdx) => {
                  return (
                    <li key={subIdx}>
                      {createIcon(sub.status)} {sub.text}
                    </li>
                  );
                })}
              </ul>
            )}
          </div>
        );
  };

  render() {
    const health = this.props.health;
    return health.health.items.map((item, idx) => {
      return this.renderChildren(item, idx);
    });
  }
}
