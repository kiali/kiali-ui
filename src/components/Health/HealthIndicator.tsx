import * as React from 'react';
import { PopoverPosition, Text, TextContent, TextVariants, Tooltip } from '@patternfly/react-core';
import { HealthDetails } from './HealthDetails';
import * as H from '../../types/Health';
import { createIcon } from './Helper';
import './Health.css';
import { PfColors } from '../Pf/PfColors';

export enum DisplayMode {
  LARGE,
  SMALL
}

interface Props {
  id: string;
  health?: H.Health;
  mode: DisplayMode;
  tooltipPlacement?: PopoverPosition;
}

interface HealthState {
  globalStatus: H.Status;
}

export class HealthIndicator extends React.PureComponent<Props, HealthState> {
  static getDerivedStateFromProps(props: Props) {
    return {
      globalStatus: props.health ? props.health.getGlobalStatus() : H.NA
    };
  }

  constructor(props: Props) {
    super(props);
    this.state = HealthIndicator.getDerivedStateFromProps(props);
  }

  render() {
    if (this.props.health) {
      if (this.props.mode === DisplayMode.SMALL) {
        return this.renderSmall(this.props.health);
      } else {
        return this.renderLarge(this.props.health);
      }
    }
    return <span />;
  }

  renderSmall(health: H.Health) {
    return this.renderPopover(health, createIcon(this.state.globalStatus, 'sm'));
  }

  renderLarge(health: H.Health) {
    const spanStyle: React.CSSProperties = {
      color: this.state.globalStatus.color,
      fontWeight: 'bold',
      position: 'relative',
      top: -9,
      left: 10
    };
    return (
      <>
        {createIcon(this.state.globalStatus, 'lg')}
        <span style={spanStyle}>{this.state.globalStatus.name}</span>
        <br />
        <br />
        <HealthDetails health={health} />
      </>
    );
  }
  renderHealthTooltip(health: H.Health) {
    return (
      <TextContent style={{ color: PfColors.White }}>
        <Text component={TextVariants.h2}>{this.state.globalStatus.name}</Text>
        <HealthDetails health={health} />
      </TextContent>
    );
  }

  renderPopover(health: H.Health, icon: JSX.Element) {
    return (
      <Tooltip
        aria-label={'Health indicator'}
        content={this.renderHealthTooltip(health)}
        position={PopoverPosition.auto}
        className={'health_indicator'}
      >
        {icon}
      </Tooltip>
    );
  }
}
