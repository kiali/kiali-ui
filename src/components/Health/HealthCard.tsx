import * as React from 'react';
import { Card, CardActions, CardBody, CardHead, CardHeader, Stack, StackItem, Title } from '@patternfly/react-core';
import { DisplayMode, HealthIndicator } from './HealthIndicator';
import { style } from 'typestyle';
import * as H from '../../types/Health';

type Props = {
  name: string;
  health?: H.Health;
};

const titleStyle = style({
  margin: '15px 0 11px 0'
});

class HealthCard extends React.PureComponent<Props> {
  render() {
    return (
      <Card>
        <CardHead>
          <CardActions />
          <CardHeader>
            <Title style={{ float: 'left' }} headingLevel="h3" size="2xl">
              Health
            </Title>
          </CardHeader>
        </CardHead>
        <CardBody>
          <Stack className={'stack_service_details'}>
            <StackItem id={'health'}>
              <Title headingLevel="h3" size="lg" className={titleStyle}>
                Overall Health
              </Title>
              <HealthIndicator id={this.props.name} health={this.props.health} mode={DisplayMode.LARGE} />
            </StackItem>
          </Stack>
        </CardBody>
      </Card>
    );
  }
}

export default HealthCard;
