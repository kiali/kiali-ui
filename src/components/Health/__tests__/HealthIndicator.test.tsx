import * as React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { HealthIndicator, DisplayMode } from '../HealthIndicator';
import { AppHealth } from '../../../types/Health';
import { PFAlertColor } from 'components/Pf/PfColors';

describe('HealthIndicator', () => {
  it('renders when empty', () => {
    // SMALL
    let wrapper = shallow(<HealthIndicator id="svc" mode={DisplayMode.SMALL} />);
    expect(wrapper.html()).not.toContain('pficon');

    // LARGE
    wrapper = shallow(<HealthIndicator id="svc" mode={DisplayMode.LARGE} />);
    expect(wrapper.html()).not.toContain('pficon');
  });

  it('renders healthy', () => {
    const health = new AppHealth(
      [
        { name: 'A', availableReplicas: 1, currentReplicas: 1, desiredReplicas: 1 },
        { name: 'B', availableReplicas: 2, currentReplicas: 2, desiredReplicas: 2 }
      ],
      { errorRatio: -1, inboundErrorRatio: -1, outboundErrorRatio: -1 },
      { rateInterval: 600, hasSidecar: true }
    );

    // SMALL
    let wrapper = shallow(<HealthIndicator id="svc" health={health} mode={DisplayMode.SMALL} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    let html = wrapper.html();
    expect(html).toContain(PFAlertColor.Success);

    // LARGE
    wrapper = shallow(<HealthIndicator id="svc" health={health} mode={DisplayMode.LARGE} />);
    html = wrapper.html();
    expect(html).toContain(PFAlertColor.Success);
  });

  it('renders workloads degraded', () => {
    const health = new AppHealth(
      [
        { name: 'A', availableReplicas: 1, currentReplicas: 1, desiredReplicas: 10 },
        { name: 'B', availableReplicas: 2, currentReplicas: 2, desiredReplicas: 2 }
      ],
      { errorRatio: -1, inboundErrorRatio: -1, outboundErrorRatio: -1 },
      { rateInterval: 600, hasSidecar: true }
    );

    // SMALL
    let wrapper = shallow(<HealthIndicator id="svc" health={health} mode={DisplayMode.SMALL} />);
    let html = wrapper.html();
    expect(html).toContain(PFAlertColor.Warning);

    // LARGE
    wrapper = shallow(<HealthIndicator id="svc" health={health} mode={DisplayMode.LARGE} />);
    html = wrapper.html();
    expect(html).toContain(PFAlertColor.Warning);
    expect(html).toContain('1 / 10');
  });

  it('renders some scaled down workload', () => {
    const health = new AppHealth(
      [
        { name: 'A', availableReplicas: 0, currentReplicas: 0, desiredReplicas: 0 },
        { name: 'B', availableReplicas: 2, currentReplicas: 2, desiredReplicas: 2 }
      ],
      { errorRatio: -1, inboundErrorRatio: -1, outboundErrorRatio: -1 },
      { rateInterval: 600, hasSidecar: true }
    );

    // SMALL
    let wrapper = shallow(<HealthIndicator id="svc" health={health} mode={DisplayMode.SMALL} />);
    let html = wrapper.html();
    expect(html).toContain(PFAlertColor.Success);

    // LARGE
    wrapper = shallow(<HealthIndicator id="svc" health={health} mode={DisplayMode.LARGE} />);
    html = wrapper.html();
    expect(html).toContain(PFAlertColor.Success);
    expect(html).toContain('0 / 0');
  });

  it('renders all workloads down', () => {
    const health = new AppHealth(
      [
        { name: 'A', availableReplicas: 0, currentReplicas: 0, desiredReplicas: 0 },
        { name: 'B', availableReplicas: 0, currentReplicas: 0, desiredReplicas: 0 }
      ],
      { errorRatio: -1, inboundErrorRatio: -1, outboundErrorRatio: -1 },
      { rateInterval: 600, hasSidecar: true }
    );

    // SMALL
    let wrapper = shallow(<HealthIndicator id="svc" health={health} mode={DisplayMode.SMALL} />);
    let html = wrapper.html();
    expect(html).toContain(PFAlertColor.Danger);

    // LARGE
    wrapper = shallow(<HealthIndicator id="svc" health={health} mode={DisplayMode.LARGE} />);
    html = wrapper.html();
    expect(html).toContain(PFAlertColor.Danger);
  });

  it('renders error rate failure', () => {
    const health = new AppHealth(
      [{ name: 'A', availableReplicas: 1, currentReplicas: 1, desiredReplicas: 1 }],
      { errorRatio: 0.3, inboundErrorRatio: 0.1, outboundErrorRatio: 0.2 },
      { rateInterval: 600, hasSidecar: true }
    );

    // SMALL
    let wrapper = shallow(<HealthIndicator id="svc" health={health} mode={DisplayMode.SMALL} />);
    let html = wrapper.html();
    expect(html).toContain(PFAlertColor.Danger);

    // LARGE
    wrapper = shallow(<HealthIndicator id="svc" health={health} mode={DisplayMode.LARGE} />);
    html = wrapper.html();
    expect(html).toContain(PFAlertColor.Danger);
    expect(html).toContain('Outbound: 20.00%');
    expect(html).toContain('Inbound: 10.00%');
  });
});
