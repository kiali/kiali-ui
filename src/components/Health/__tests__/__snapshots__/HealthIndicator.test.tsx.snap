// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HealthIndicator renders healthy 1`] = `
<OverlayTrigger
  defaultOverlayShown={false}
  overlay={
    <Popover
      bsClass="popover"
      id="svc-health-tooltip"
      placement="right"
      title="Healthy"
    >
      <HealthDetails
        health={
          AppHealth {
            "items": Array [
              Object {
                "children": Array [
                  Object {
                    "status": Object {
                      "class": "icon-healthy",
                      "color": "#3f9c35",
                      "icon": [Function],
                      "name": "Healthy",
                      "priority": 1,
                    },
                    "text": "A: 1 / 1",
                  },
                  Object {
                    "status": Object {
                      "class": "icon-healthy",
                      "color": "#3f9c35",
                      "icon": [Function],
                      "name": "Healthy",
                      "priority": 1,
                    },
                    "text": "B: 2 / 2",
                  },
                ],
                "status": Object {
                  "class": "icon-healthy",
                  "color": "#3f9c35",
                  "icon": [Function],
                  "name": "Healthy",
                  "priority": 1,
                },
                "title": "Pods Status",
              },
              Object {
                "children": Array [
                  Object {
                    "status": Object {
                      "class": "icon-na",
                      "color": "#72767b",
                      "icon": [Function],
                      "name": "No health information",
                      "priority": 0,
                    },
                    "text": "Inbound: No requests",
                  },
                  Object {
                    "status": Object {
                      "class": "icon-na",
                      "color": "#72767b",
                      "icon": [Function],
                      "name": "No health information",
                      "priority": 0,
                    },
                    "text": "Outbound: No requests",
                  },
                ],
                "status": Object {
                  "class": "icon-na",
                  "color": "#72767b",
                  "icon": [Function],
                  "name": "No health information",
                  "priority": 0,
                },
                "title": "Error Rate over last 10m",
              },
            ],
            "requests": Object {
              "errorRatio": -1,
              "inboundErrorRatio": -1,
              "outboundErrorRatio": -1,
            },
          }
        }
      />
    </Popover>
  }
  placement="right"
  rootClose={false}
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <OkIcon
    className="icon-healthy"
    color="#3f9c35"
    noVerticalAlign={false}
    size="sm"
    title={null}
  />
</OverlayTrigger>
`;
